### PR TITLE
[PORT] Brings UpdatePaths uptodate

### PR DESCRIPTION
**Pull request by mystery3525**
## About The Pull Request

Ports the following:
- https://github.com/tgstation/tgstation/pull/75892
- https://github.com/tgstation/tgstation/pull/70806
- https://github.com/tgstation/tgstation/pull/68436
- https://github.com/tgstation/tgstation/pull/67489
- https://github.com/tgstation/tgstation/pull/67241 (in principle)
- https://github.com/tgstation/tgstation/pull/66916
- https://github.com/tgstation/tgstation/pull/66506
- https://github.com/tgstation/tgstation/pull/66469
- https://github.com/tgstation/tgstation/pull/62595
- https://github.com/tgstation/tgstation/pull/56573 (one line change)

Various fixes and improvements to update paths over the years, gives us really good tooling for massive refactors and ports.

## Why It's Good For The Game

This allows us to do a wide variety of changes to all maps, all at once, instead of having to manually map change for simple repaths, etc. Also fixes UpdatePaths as it didn't work at all before.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>
not really sure how I can test this?

</details>

## Changelog
:cl: AnturK, esainane, MrMelbert, Son-of-Space, san7890, Ghommie, ported by mystery3525
fix: Various fixes to the UpdatePaths tool
fix: UpdatePaths will now handle lists that contain strings.
code: Adds support for simple subtype repathing, outright deletion,  for UpdatePaths scripts
code: Orginizes UpdatePaths folder and adds a well-documented readme.md
/:cl:

### CODE REVIEW

**Comment by Tsar-Salat** [comment]
oh thank god

**Comment by mystery3525** [comment]
image

File: tools/mapmerge2/dmm.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def from_bytes(bytes):
 
     def to_file(self, fname, *, tgm = True):
         self._presave_checks()
-        with open(fname, 'w', newline='\n', encoding=ENCODING) as f:
+        with open(fname, 'w', newline='\r\n', encoding=ENCODING) as f:
```
Comment (itsmeow) [inline tools/mapmerge2/dmm.py:34]: why?

Comment (mystery3525) [reply tools/mapmerge2/dmm.py:34]: image


image

Comment (mystery3525) [reply tools/mapmerge2/dmm.py:34]: image

**Review by itsmeow — APPROVED** [review]

**Comment by itsmeow** [comment]
@Crossedfall

File: tools/UpdatePaths/__main__.py [hunk 2]
```diff
@@ -87,18 +93,29 @@ def replace_def(match):
...
         for new_path, new_props in new_paths:
             if new_path == "@OLD":
                 out = match.group('path')
+            elif new_path == "@DELETE":
+                if verbose:
+                    print("Deleting match : {0}".format(match.group(0)))
+                return [None]
+            elif new_path.endswith("/@SUBTYPES"):
+                path_start = new_path[:-len("/@SUBTYPES")]
+                out = path_start + match.group('subtype')
```
Comment (Crossedfall) [inline tools/UpdatePaths/__main__.py]:
```suggestion
                path_start = new_path[ :-len("/@SUBTYPES")]
                out = path_start + match.group("subtype")
```

File: tools/UpdatePaths/__main__.py [hunk 3]
```diff
@@ -126,21 +142,26 @@ def get_result(element):
             return [element]
 
     bad_keys = {}
+    modified_keys = []
     keys = list(dmm_data.dictionary.keys())
     for definition_key in keys:
         def_value = dmm_data.dictionary[definition_key]
-        new_value = tuple(y for x in def_value for y in get_result(x))
+        new_value = tuple(y for x in def_value for y in get_result(x) if y != None)
```
Comment (Crossedfall) [inline tools/UpdatePaths/__main__.py]:
```suggestion
        new_value = tuple(y for x in def_value for y in get_result(x) if y is not None)
```

File: tools/mapmerge2/dmm.py (hunk 1 above)
Comment (Crossedfall) [inline tools/mapmerge2/dmm.py:34]: Mmm, I don't like this, but it's something we can fix in a separate PR.

File: tools/mapmerge2/dmm.py [hunk 4]
```diff
@@ -83,6 +83,14 @@ def reassign_bad_keys(self, bad_keys):
             # reassign the grid entries which used the old key
             self.grid[k] = bad_keys.get(v, v)
 
+    def remove_unused_keys(self, modified_keys = None):
+        unused_keys = list(set(modified_keys)) if modified_keys is not None else self.dictionary.keys()
+        for key in self.grid.values():
+            if key in unused_keys:
+                unused_keys.remove(key)
+        for key in unused_keys:
+            del self.dictionary[key]
```
Comment (Crossedfall) [inline tools/mapmerge2/dmm.py]:
```suggestion
        for key in unused_keys:
            self.dictionary.pop(key, None)
```

This is safer and a bit faster

**Review by Crossedfall — CHANGES_REQUESTED** [review]
Wow, this file is pretty terrible. For the most part your changes look fine, I made some minor suggestions. We should probably look at normalizing *.dmm files. I don't recall why we skipped those.

**Review by Crossedfall — APPROVED** [review]

**Pull request merged**